Runtime/C#/Runtime/BebopWriteResult.cs

```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Bebop.Runtime
+{
+    /// <summary>
+    /// Details the result of attempting to encode a record to a buffer
+    /// </summary>
+    public readonly struct BebopWriteResult
+    {
+        /// <summary></summary>
+        public BebopWriteResult(WriteResult result, int written, byte[] buffer)
+        {
+            Result = result;
+            Written = written;
+            Buffer = buffer;
+        }
+
+        /// <summary>
+        /// The result of encoding. If this succeeded, or a specific failure type
+        /// </summary>
+        public WriteResult Result { get; }
+
+        /// <summary>
+        /// The amount of bytes written to the buffer
+        /// </summary>
+        public int Written { get; }
+
+        /// <summary>
+        /// The final instance of the buffer written to. If Bebop Writer was given a buffer to start with, but had to grow it to fit the fill serialized message, this will be a different instance than what was passed in.
+        /// </summary>
+        public byte[] Buffer { get; }
```

Comment:
A few things:

1) [Properties should not return arrays](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1819). 

2) If `BebopWriter` is given a buffer that is too small to contain the encoded record, then we should treat this as an error and return `BufferTooSmall` rather than attempting to grow the array and returning `SuccessButGrewBuffer`. It is the responsibility of developers to ensure their buffers are the right size. 

With that change you can remove the `Buffer` property from this struct and upon a successful encoding the buffer that was passed into `EncodeIntoBuffer` will be the same one that contains the encoded bytes.

Reply:
I prefer what you suggested, but this one will be difficult. Since `void BebopWriter.GrowBy()` is called each time a field is written to the buffer, even if it doesn't need to grow, I'm worried we'd have to change it to return something like `EncodeResult` on each call, then add an if check around it to know if we should keep encoding, or return immediatly.

How do you feel about throwing an exception inside `BebopWriter.GrowBy()` when that instance of `BebopWriter` can't grow an input buffer? I haven't benchmarked it yet, but here's what I'm thinking it would end up looking like:
```cs
private void GrowBy(int amount)
{
    if ((Length & 0xC0000000) != 0)
    {
        throw new BebopViewException("A Bebop View cannot grow beyond 2 gigabytes.");
    }
    if (Length + amount > _buffer.Length)
    {
        // START NEW CODE
        if (_isInputBuffer) // Note: _isInputBuffer is set by the constructor
        {
            throw new BebopViewException("A Bebop View cannot grow a buffer that was input during View creation.");
        }
        // END NEW CODE

        //snip: grow buffer
    }
    Length += amount;
}
```

Reply:
I ended up implementing something very similar to this approach. Ran the same benchmarks and didn't notice any performance regression.